### PR TITLE
WIP - Vendor ARA client libs

### PR DESCRIPTION
**Pull request by Thulium-Drake**
:dragon_face: ! HERE BE DRAGONS ! :dragon_face: 

So this is a highly experimental change, I mainly made the PR for proof of concept purposes.

The idea behind is is the same as what the Foreman guys did last october with https://github.com/theforeman/foreman-ansible-modules/commit/0eb76b8b5ab1bda9d82e5209b9bd2832759ae0e1

This change allows systems without any ARA package installed from Pypi to still talk to ARA API's:

```
root@ansible-dev (vendor_ara_client):ara$ pip3 list | grep ara
paramiko                 2.7.1
python-saharaclient      2.0.0
root@ansible-dev (vendor_ara_client):ara$ pip3 list --user | grep ara
root@ansible-dev (vendor_ara_client):ara$ 
```

Required ```ansible.cfg``` snippet
```
callback_whitelist = community.general.yaml, recordsansible.ara.ara_default
```

image

So the question is, do we want to go this route?

### CODE REVIEW

**Comment by dmsimard** [comment]
Hey @Thulium-Drake and thanks for putting this together, it was worth it and helpful to get a good picture.

> So the question is, do we want to go this route?

I talked a bit about this in the issue you linked but in short: I don't think so because the pros from having the API, plugins, CLI and clients bundled together in a single package outweighs the pros of having it inside the collection.

In other words, running ``pip install ara`` gets you the Ansible plugins, the [CLI](https://ara.readthedocs.io/en/latest/cli.html) and the means to talk to a remote API server by way of using the client libraries that are included.

I feel the purpose of the collection should be more about installing and configuring ara than it is about consuming (or using) ara, if that makes sense.

**Comment by Thulium-Drake** [comment]
Hi @dmsimard ,

You are quite correct, splitting up seems like a bad idea. And you're right w.r.t. the pip package also contains the ARA callback (D'Oh!). I had a checkout of the git repo on my system and symlinked to it, which is prone to errors (like me forgetting to update it for a while).

I understand your PoV in this regard, this collection is more about 'getting it running' then it is using it. But I can't completely ignore the need _I_ feel that getting the ARA client on a controller can be more integrated. Which also removes the need to do all kinds of virtualenv magic on controllers (when installing ara from pip on a RHEL system, you need to put it in a virtualenv which will basically break ansible's 'system' modules on that system) 

And I'm trying to think of a means where we can make a nice hybrid solution that is not as hacky as what I did above. Perhaps pull in the then-current version of the client libs before tagging a new release?

Updating the ```import```s to be compatible with both paths isn't much of a problem, I just was a bit lazy when doing this, as I wanted it to just work :innocent: 

And I will also just admit I don't like pip because it lacks a proper update mechanism :)

**Comment by dmsimard** [comment]
@Thulium-Drake I think the takeaway here is that installing ara from pip works well when ansible is also installed from pip but when ansible is installed from distribution packages, not so much (or it's complicated). I get and understand that.

I don't have any great ideas right now but at a minimum it'd be appropriate to write this down in an issue in the upstream ara repository so we can take care of it eventually.

We could go down the path of documenting things like setting ``PYTHONPATH`` and such (like AWX and Tower venvs might) but that feels hacky as well and I'd rather keep things simple if we can help it.

It's not something that I have been looking forward to support because, well, it's harder and more complex to support than pointing users down the simple and easy way of getting everything from pypi.

**Comment by Thulium-Drake** [comment]
I agree on the ```PYTHONPATH``` being hacky. Come to think of it, the best solution might be to 'just' package them. I have a script that allows me to make ```deb``` packages, but those hardly qualify for broader usage.

But that will basically be the fix for virtualenv magic :)

Shall I make an issue for packaging it? (I feel that RPMs will fix most issues, as in my experience on Debian this works fine)

**Comment by dmsimard** [comment]
> Shall I make an issue for packaging it?

It depends on what you think we should do:

- I maintain an official ara package for fedora: https://koji.fedoraproject.org/koji/packageinfo?packageID=24394
- arrfab maintains a package in the CentOS configuration management SIG: https://cbs.centos.org/koji/packageinfo?packageID=8167
- There is also a package for Debian but it seems to be lagging a bit behind right now: https://tracker.debian.org/pkg/python-ara
- There is even a package for opensuse: https://build.opensuse.org/package/show/devel:languages:python/python-ara

**Comment by Thulium-Drake** [comment]
> > Shall I make an issue for packaging it?
Oh wow, I genuinely missed most of these!
> 
> It depends on what you think we should do:
> 
>     * I maintain an official ara package for fedora: https://koji.fedoraproject.org/koji/packageinfo?packageID=24394
Knew that :smile: 
> 
>     * arrfab maintains a package in the CentOS configuration management SIG: https://cbs.centos.org/koji/packageinfo?packageID=8167
> 
This is really nice! I'm thinking, it's not a distribution package. We could add a flag to the role to let it try to install packages named ```ara``` and ```ara-server``` anyway, provided that the user configures the required extra repositories (I can see this happening via Foreman/Satellite when this repo is synced to it)
>     * There is also a package for Debian but it seems to be lagging a bit behind right now: https://tracker.debian.org/pkg/python-ara
> 
Aww, that's a shame! I'll try contacting the guy to at least show my appreciation for it and ask him what needs to be done to make this an actual thing in Debian.
>     * There is even a package for opensuse: https://build.opensuse.org/package/show/devel:languages:python/python-ara
Nice!

So, bottom line, packaging's sorted for most cases :-) except that Fedora's the only distro that has it in it's repo's.

**Comment by Thulium-Drake** [comment]
Closing this as discussed.
